### PR TITLE
dns_cache: synchronize access with mutex

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
In https://github.com/RIOT-OS/RIOT/pull/18329#discussion_r924622627, @benpicco realized, that synchronization of cache access is missing for `dns_cache`. This adds the missing synchronization.

### Testing procedure
`tests/unittests` should still pass.

### Issues/PRs references
Follow-up on #18318.

### CODE REVIEW

**Comment by miri64** [comment]
@chrysn this _is_ a bug fix, right? ;-)

File: sys/net/application_layer/dns/cache.c [hunk 1]
```diff
@@ -137,10 +140,12 @@ int dns_cache_query(const char *domain_name, void *addr_out, int family)
         if (cache[i].hash == hash && (!addr_len || addr_len == _get_len(i))) {
             DEBUG("dns_cache[%u] hit\n", i);
             memcpy(addr_out, &cache[i].addr, _get_len(i));
+            mutex_unlock(&cache_mutex);
             return _get_len(i);
```
Comment (benpicco) [inline sys/net/application_layer/dns/cache.c:144]: I smell a potential race condition

Comment (benpicco) [inline sys/net/application_layer/dns/cache.c:144]:
```suggestion
            uint8_t len = _get_len(i)
            mutex_unlock(&cache_mutex);
            return len;
```

Comment (miri64) [reply sys/net/application_layer/dns/cache.c]: Fixed with the suggestion provided in https://github.com/RIOT-OS/RIOT/pull/18330#discussion_r924690579

**Review by benpicco — APPROVED** [review]
Thank you for the fix!

**Comment by miri64** [comment]
As this is more or less a bug fix, should this be backported @chrysn? (even though it did not show up in the release-spec tests as an issue ;-))

**Comment by miri64** [comment]
Took the opportunity of the syntax error and cleaned up the code. ;-)

**Comment by chrysn** [comment]
Seems GitHub swallowed my response mail -- so here again:

I guess I opened that can of worms myself with the late addition of the the cache unification -- but Wer die Dose öffnet, muss sie auch auslöffeln. (Sorry, doesn't translate well.) Please file for backporting.

**Comment by miri64** [comment]
Backport provided in #18342

**Comment by chrysn** [comment]
I guess I opened that can of worms myself with the late addition of the the cache unification -- but Wer die Dose öffnet, muss sie auch auslöffeln. (Sorry, doesn't translate well.) Please file for backporting.

**Comment by miri64** [comment]
> I guess I opened that can of worms myself with the late addition of the the cache unification -- but Wer die Dose öffnet, muss sie auch auslöffeln. (Sorry, doesn't translate well.) Please file for backporting

Mh? The backport was merged to 2022.07 in https://github.com/RIOT-OS/RIOT/pull/18342

**Comment by miri64** [comment]
Ah apparently the mail you wrote about in https://github.com/RIOT-OS/RIOT/pull/18330#issuecomment-1189985632 finally came through :sweat_smile:

**Pull request merged**